test/integration/xds_config_tracker_integration_test.cc

```diff
@@ -0,0 +1,248 @@
+    // we need a dummy data plane cluster.
+    setUpstreamCount(1);
+    setUpstreamProtocol(Http::CodecType::HTTP2);
+    HttpIntegrationTest::initialize();
+    registerTestServerPorts({});
+    initial_xds_update_ = test_server_->counter("test_xds_tracker.on_config_accepted")->value();
+    initial_xds_reject_ = test_server_->counter("test_xds_tracker.on_config_rejected")->value();
+  }
+
+  void acceptXdsConnection() {
+    createXdsConnection();
+    AssertionResult result = xds_connection_->waitForNewStream(*dispatcher_, xds_stream_);
+    RELEASE_ASSERT(result, result.message());
+    xds_stream_->startGrpcStream();
+  }
+
+  uint32_t initial_xds_update_{0};
+  uint32_t initial_xds_reject_{0};
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersionsClientType, XdsConfigTrackerIntegrationTest,
+                         DELTA_SOTW_UNIFIED_GRPC_CLIENT_INTEGRATION_PARAMS);
+
+TEST_P(XdsConfigTrackerIntegrationTest, XdsConfigTrackerSuccessCount) {
+  TestXdsConfigTrackerFactory factory;
+  Registry::InjectFactory<Config::XdsConfigTrackerFactory> registered(factory);
+
+  initialize();
+  acceptXdsConnection();
+
+  EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Runtime, "", {"some_rtds_layer"},
+                                      {"some_rtds_layer"}, {}, true));
+  auto some_rtds_layer = TestUtility::parseYaml<envoy::service::runtime::v3::Runtime>(R"EOF(
+    name: some_rtds_layer
+    layer:
+      foo: bar
+      baz: meh
+  )EOF");
+  sendDiscoveryResponse<envoy::service::runtime::v3::Runtime>(
+      Config::TypeUrl::get().Runtime, {some_rtds_layer}, {some_rtds_layer}, {}, "1");
+  test_server_->waitForCounterEq("test_xds_tracker.on_config_accepted", initial_xds_update_ + 1);
+  EXPECT_EQ(test_server_->counter("test_xds_tracker.on_config_accepted")->value(),
+            initial_xds_update_ + 1);
+}
+
+TEST_P(XdsConfigTrackerIntegrationTest, XdsConfigTrackerFailureCount) {
+  TestXdsConfigTrackerFactory factory;
+  Registry::InjectFactory<Config::XdsConfigTrackerFactory> registered(factory);
+
+  initialize();
+  acceptXdsConnection();
+
+  EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Runtime, "", {"some_rtds_layer"},
+                                      {"some_rtds_layer"}, {}, true));
+  auto some_rtds_layer = TestUtility::parseYaml<envoy::service::runtime::v3::Runtime>(R"EOF(
+    name: different_rtds_layer
+    layer:
+      foo: bar
+      baz: meh
+  )EOF");
+
+  auto route_config = TestUtility::parseYaml<envoy::config::route::v3::RouteConfiguration>(R"EOF(
+    name: my_route
+    vhds:
+      config_source:
+        resource_api_version: V3
+        api_config_source:
+          api_type: GRPC
+          transport_api_version: V3
+          grpc_services:
+            envoy_grpc:
+              cluster_name: xds_cluster
+    )EOF");
+
+  FakeStream* stream = xds_stream_.get();
+  envoy::service::discovery::v3::DiscoveryResponse discovery_response;
```

Comment:
This seems odd, for non-SotW responses. Should this cover both Delta and SotW?

Reply:
My bad. Yes, we need to simulate both, there is an existing template method but is type strict, and this simulate an exception scenario.